version: '3.9'

services:
  flask_api:
    build: flask_app
    restart: always
    container_name: flask_app
    volumes:
      - ./flask_app/src:/usr/src/app
    env_file:
      - src/core/.env
    command: python pywsgi.py
    expose:
      - "5000"
    networks:
      - api
    depends_on:
      database:
        condition: service_healthy

  database:
    image: postgres:13-alpine3.16
    restart: always
    container_name: postgres
    expose:
      - "5432"
    env_file:
      - database/.env
    volumes:
      - ./database/init_data/:/docker-entrypoint-initdb.d/
      - ./database/data:/var/lib/postgresql/data/
    networks:
      - api
    healthcheck:
      test: [ "CMD-SHELL", "su postgres", "&&", "pg_isready" ]
      interval: 5s
      timeout: 5s
      retries: 10

  redis:
    image: redis:6.2.6
    restart: always
    container_name: redis_cashe
    expose:
      - "6379"
    command: redis-server --save 20 1 --loglevel warning
    volumes:
      - cache:/data
    networks:
      - api
    healthcheck:
      test: [ "CMD", "redis-cli", "ping" ]
      interval: 5s
      timeout: 30s
      retries: 50


  nginx:
    image: nginx:1.21.6
    restart: always
    container_name: nginx
    volumes:
      - ./nginx/nginx.conf:/etc/nginx/nginx.conf:ro
      - ./nginx/configs:/etc/nginx/conf.d:ro
    ports:
      - "80:80"
    networks:
      - api
    depends_on:
      - flask_api

volumes:
  cache:
    driver: local

networks:
  api:
    driver: bridge